tests/test_stdiomgr_base.py

```diff
@@ -115,6 +116,22 @@ def test_repeated_use():
         test_capture_stderr()
 
 
+def test_noop():
+    """Confirm state is restored after context."""
+    real_sys_stdio = (sys.stdin, sys.stdout, sys.stderr)
+    stdio_mgr()
+    assert (sys.stdin, sys.stdout, sys.stderr) == real_sys_stdio
```

Comment:
Definitely good tests to add.

Should this (and also in `test_exception`) be a stricter identity check, though? I don't know how equality is implemented on `TextIOWrapper`, and if there are subtle gotchas.

Also, I'd prefer to assert-report the actual streams on assert failure.

Am picturing replacing this current `assert` line with something like:

```
final_sys_stdio = (sys.stdin, sys.stdout, sys.stderr)
assert all(f is r for f, r in zip(final_sys_stdio, real_sys_stdio)), final_sys_stdio
```

Reply:
```
$ git checkout master src
$ git diff
```
```diff
diff --git a/tox.ini b/tox.ini
index 57af8b8..5a986dd 100644
--- a/tox.ini
+++ b/tox.ini
@@ -10,7 +10,7 @@ envlist=

 [testenv]
 commands=
-    pytest
+    pytest -v
 deps=
     attrs_17_1:     attrs==17.1
     attrs_17_2:     attrs==17.2
```
```
$ python3 -m tox -e py37-attrs_latest
....
_____________________________________________________________________________ test_exception _____________________________________________________________________________

    def test_exception():
        """Confirm state is restored after an exception during context."""
        real_sys_stdio = (sys.stdin, sys.stdout, sys.stderr)
        with pytest.raises(ZeroDivisionError):
            with stdio_mgr() as (i, o, e):
                1 / 0
>       assert (sys.stdin, sys.stdout, sys.stderr) == real_sys_stdio
E       AssertionError: assert (SafeCloseTee...ding='utf-8'>) == (<_pytest.capt...7f1769e83470>)
E         At index 0 diff: SafeCloseTeeStdin(tee=<_io.TextIOWrapper encoding='utf-8'>, init_text='', _encoding='utf-8') != <_pytest.capture.DontReadFromInput object at 0x7f1769e83358>
E         Full diff:
E         - (SafeCloseTeeStdin(tee=<_io.TextIOWrapper encoding='utf-8'>, init_text='', _encoding='utf-8'),
E         -  <_io.TextIOWrapper encoding='utf-8'>,
E         -  <_io.TextIOWrapper encoding='utf-8'>)
E         + (<_pytest.capture.DontReadFromInput object at 0x7f1769e83358>,
E         +  <_pytest.capture.EncodedFile object at 0x7f1769e833c8>,...
E
E         ...Full output truncated (2 lines hidden), use '-vv' to show

tests/test_stdiomgr_base.py:132: AssertionError
```

vs, with your proposed change.

```
E       AssertionError: (SafeCloseTeeStdin(tee=<_io.TextIOWrapper encoding='utf-8'>, init_text='', _encoding='utf-8'), <_io.TextIOWrapper encoding='utf-8'>, <_io.TextIOWrapper encoding='utf-8'>)
E       assert False
E        +  where False = all(<generator object test_exception.<locals>.<genexpr> at 0x7f2b59b4ab10>)
```
Adding list may help, but I have to zip off to read my son a bed time story...

As for TextIOWrapper, it says its __eq__ is object.__eq__ , so there is no magic equivalence happening there.

Reply:
LOL. My "clever idea" just made things even less clear.

Looks like the dressing pytest puts on the stdio does funny things to the standard regardless... it's simpler the way you originally wrote it, works, and gives a measure of debugging detail. Cancel requested change. 🚫